### PR TITLE
[Architecture] Refactor FusedMoE

### PR DESCRIPTION
**Pull request by hlu1**
## Description

Refactor FusedMoE
- Move all moe related code to `fused_moe/`
- Move all routing methods to `routing.py`. No other code change is made to routing.
- Move all quant changes to `quantization.py` and apply the same refactoring strategy in `linear.py`
- Separate out trtllm-gen code into class `TRTLLMGenFusedMoE`
- More refactoring is necessary to separate the different forward paths in `CutlassFuseMoE`. 
- Remove the hack in FusedCutlass for MTP. There was a hack to force the MTP layer to use Cutlass backend when it's nvfp4 and moe_backend=="TRTLLM" and is_mtp. This is replaced with `DeepseekV3DecoderLayer._get_decoder_layer_quant_config`, which checks if the a particular decoder layer is excluded from quantization.

## Test Coverage

Manual test
- Deepseek min-latency benchmark

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--disable-fail-fast --skip-test --stage-list "A10-1, xxx" --gpu-type "A30, H100_PCIe" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-[Post-Merge]-1, xxx"]`

Launch build/test pipelines. All previously running jobs will be killed.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests. Will also run L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-[Post-Merge]-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-[Post-Merge]-1, xxx".

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>

### CODE REVIEW

File: tensorrt_llm/_torch/models/modeling_deepseekv3.py [hunk 1]
```diff
@@ -365,7 +365,7 @@ def __init__(self,
             fuse_routing_kernel=True,
             apply_routing=False,
             moe_backend=model_config.moe_backend)
-        self.experts = FusedMoE(
+        self.experts = get_moe_cls(model_config)(
```
Comment (yuxianq) [inline tensorrt_llm/_torch/models/modeling_deepseekv3.py:368]: Can we use `create_fused_moe(num_experts=self.num_experts, ...)` like `create_attention` instead of get moe class first?

File: tensorrt_llm/_torch/modules/fused_moe/interface.py [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+from ...model_config import ModelConfig
+from .routing import BaseMoeRoutingMethod
+
+
+class MoEWeightLoadingMode(Enum):
+    VANILLA = 0
+    FUSED_GATE_UP_PROJ = 1
+
+
+class FusedMoE(nn.Module):
```
Comment (yuxianq) [inline tensorrt_llm/_torch/modules/fused_moe/interface.py:17]: Should we use `MoE` instead of `FusedMoE` since some implementations like `VanillaMoE` do not fuse at all.

File: tensorrt_llm/_torch/modules/fused_moe/quantization.py [hunk 3]
```diff
@@ -0,0 +1,1181 @@
+import threading
+from abc import ABC, abstractmethod
+from typing import Dict, List, NamedTuple, override
+
+import torch
+from torch import nn
+
+from tensorrt_llm._utils import get_sm_version
+
+from ....quantization.utils.fp4_utils import (
```
Comment (yuxianq) [inline tensorrt_llm/_torch/modules/fused_moe/quantization.py:10]: Use absolute import instead since it is out of _torch package.

File: tensorrt_llm/_torch/modules/fused_moe/interface.py (hunk 2 above)
Comment (hlu1) [reply tensorrt_llm/_torch/modules/fused_moe/interface.py:17]: OK.

File: tensorrt_llm/_torch/models/modeling_deepseekv3.py (hunk 1 above)
Comment (hlu1) [reply tensorrt_llm/_torch/models/modeling_deepseekv3.py]: Done

File: tensorrt_llm/_torch/modules/fused_moe/fused_moe_cutlass.py [hunk 4]
```diff
@@ -0,0 +1,632 @@
...
+    def __init__(
+        self,
+        *,
+        routing_method: BaseMoeRoutingMethod,
+        num_experts: int,
+        hidden_size: int,
+        intermediate_size: int,
+        dtype: Optional[torch.dtype] = None,
+        reduce_results: bool = False,
+        model_config: ModelConfig = ModelConfig(),
```
[the remark this answers is not in the document]
Comment (hlu1) [reply tensorrt_llm/_torch/modules/fused_moe/fused_moe_cutlass.py:77]: In a follow-up PR, I can add logics to force `ModelConfig` to be readonly after construction.

[the remark this answers is not in the document]
Comment (hlu1) [reply tensorrt_llm/_torch/modules/fused_moe/fused_moe_cutlass.py:78]: https://github.com/NVIDIA/TensorRT-LLM/pull/4814

File: tensorrt_llm/_torch/modules/fused_moe/fused_moe_cutlass.py [hunk 5]
```diff
@@ -0,0 +1,640 @@
...
+
+        if self.enable_alltoall:
+            x, token_selected_slots, token_final_scales, alltoall_info = \
+                self.alltoall_prepare_maybe_dispatch(all_rank_num_tokens,
+                                                     x,
+                                                     token_selected_slots,
+                                                     token_final_scales)
+
+        x_sf = None
+        if self.has_any_quant:
```
Comment (djns99) [inline tensorrt_llm/_torch/modules/fused_moe/fused_moe_cutlass.py:320]: Maybe it doesn't belong in this refactor, but should we have a generic method/class `ActivationQuantizer` function that does this quantization and is controlled by the base quant method?
I don't exactly know how to make this correctly generic, and may require changes to the moe op

File: tensorrt_llm/_torch/modules/fused_moe/routing.py [hunk 6]
```diff
@@ -0,0 +1,264 @@
...
+                               dtype=torch.int32)
+        indices = indices.repeat(repeat_count)
+        indices = indices[:final_size]
+        balanced_indices = indices.view(router_logits.shape[0],
+                                        self.top_k).contiguous()
+
+        return balanced_indices, balanced_values
+
+
+class Qwen3MoeRoutingMethod(BaseMoeRoutingMethod):
```
Comment (djns99) [inline tensorrt_llm/_torch/modules/fused_moe/routing.py:244]: Maybe I am missing something, this is just the Renormalize method spelled less efficiently

`x = topk(softmax()); x /= x.sum()` is mathematically equivalent to `softmax(topk)`

Maybe we should add a comment explaining this isomorphism so we dont get people continually re-adding this spelling

File: tests/unittest/_torch/modules/test_fused_moe.py [hunk 7]
```diff
@@ -34,7 +34,7 @@
 
 @pytest.mark.parametrize(
     "moe_cls, dtype, experts, RoutingMethodCls",
-    product([FusedMoE, VanillaMoE], [torch.float16, torch.bfloat16],
+    product([CutlassFusedMoE, VanillaMoE], [torch.float16, torch.bfloat16],
```
Comment (djns99) [inline tests/unittest/_torch/modules/test_fused_moe.py:37]: Any chance we could make some unit tests for TRTLLM backend, I see we only have E2E model tests which are much harder to debug

**Review by djns99 — APPROVED** [review]
Thanks, this is a massive usability improvement

File: tensorrt_llm/_torch/modules/fused_moe/fused_moe_cutlass.py (hunk 5 above)
Comment (hlu1) [reply tensorrt_llm/_torch/modules/fused_moe/fused_moe_cutlass.py:320]: I thought about it and gave up the idea in the first refactor. We sometimes calls quantize and sometimes don't. It's hard to model this behavior correctly. I'll revisit this later.

File: tests/unittest/_torch/modules/test_fused_moe.py (hunk 7 above)
Comment (hlu1) [reply tests/unittest/_torch/modules/test_fused_moe.py:37]: I'll add it in a follow up PR.

File: tensorrt_llm/_torch/modules/fused_moe/routing.py (hunk 6 above)
Comment (hlu1) [reply tensorrt_llm/_torch/modules/fused_moe/routing.py:244]: I'll leave this to a follow-up PR. It'll need to rerun the CI to run qwen3 accuracy tests. This PR is blocking another 3 PRs. I need to get it in ASAP.

**Review by yuxianq — APPROVED** [review]

**Review by juney-nvidia — APPROVED** [review]
